### DISCUSSION TITLE
Make does not build deps

### DISCUSSION THREAD

**Discussion by ErikBZ** [rabbitmq/rabbitmq-server#3732] [Questions]
I'm on MacOS 11.4 and am having some issues building the rabbitmq-server from source. I installed the dependencies listed in the docs https://www.rabbitmq.com/build-server.html

 ```
 $ erl --version
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit] [dtrace]

 $ elixir --version
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit] [dtrace]

 $ make --version
GNU Make 4.3
Built for x86_64-apple-darwin20.1.0

 $ xsltproc --version
Using libxml 20904, libxslt 10129 and libexslt 817

 $ xmlto --version
xmlto version 0.0.28

 $ zip --version
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.

 $ unzip --version
caution:  both -n and -o specified; ignoring -o
UnZip 6.00 of 20 April 2009, by Info-ZIP.  Maintained by C. Spieler.  

 $ python
Python 3.8.2 (default, Jun  8 2021, 11:59:35)
[Clang 12.0.5 (clang-1205.0.22.11)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import simplejson
>>>
```

With those dependencies installed I get the following error:

```[ 11:53AM ]  [ erikzapata@Erik-Zapatas-MacBook-Pro:~/Documents/opensource/rabbitmq-server(master✗) ]
 $ make
make[1]: Entering directory '/Users/erikzapata/Documents/opensource/rabbitmq-server/deps/rabbit_common'
Error: No Makefile to build dependency /Users/erikzapata/Documents/opensource/rabbitmq-server/deps/jsx.
make[1]: *** [../../erlang.mk:4512: deps] Error 2
make[1]: Leaving directory '/Users/erikzapata/Documents/opensource/rabbitmq-server/deps/rabbit_common'
make: *** [erlang.mk:4512: deps] Error 2
```

I saw another post similar to mine https://github.com/rabbitmq/rabbitmq-server/discussions/2423 but the answer doesn't expand on which deps were configured wrong. Am I missing another dependency?

**Comment by lukebakken** [comment]
How did you clone the `rabbitmq/rabbitmq-server` repository? Could you please run these commands:

```
cd /Users/erikzapata/Documents/opensource/rabbitmq-server
git clean -fxd
rm -rf deps/*
git pull origin master
make V=2 2>&1 | tee gh-3732.txt
```

Please **attach** the `gh-3732.txt` file to your response.

**Reply by ErikBZ** [reply]
I just cloned the master branch `git clone https://github.com/rabbitmq-server`
Attached is the output of the make command
[gh-3732.txt](https://github.com/rabbitmq/rabbitmq-server/files/7540382/gh-3732.txt)

**Reply by lukebakken** [reply]
```
{"init terminating in do_boot",{undef,[{rmemo,start,[],[]},{erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,685}]},{erl_eval,exprs,5,[{file,"erl_eval.erl"},{line,123}]},{init,start_it,1,[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}
```

There is something amiss with your Erlang installation. What version of Erlang are you using and how was it installed?

Also see [these search results.](https://www.google.com/search?q=erlang+%22undef%22+%22rmemo%22+%22start%22).

Finally, if you're just trying to run RabbitMQ I **strongly suggest** using a generic unix package instead of compiling it from source - https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.9.9

**Reply by ErikBZ** [reply]
I installed it via homebrew. This is the current version I have installed:
``` $ erl --version
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit] [dtrace]
```

While trying to grab some more info I found that some dependencies build when I skip rabbit_common, by omitting it from the Makefile's DEPS env.

```
 $ make
make[1]: Entering directory '/Users/erikzapata/Documents/opensource/rabbitmq/deps/rabbit'
 DEP    ranch (2.1.0)
 DEP    ra (main)
 DEP    sysmon_handler (1.3.0)
No beam files found.
Recompile: src/rebar
Recompile: src/rebar_abnfc_compiler
Recompile: src/rebar_app_utils
Recompile: src/rebar_appups
Recompile: src/rebar_asn1_compiler
Recompile: src/rebar_base_compiler
Recompile: src/rebar_cleaner
Recompile: src/rebar_config
Recompile: src/rebar_core
Recompile: src/rebar_cover_utils
Recompile: src/rebar_ct
```

Eventually I hit the same error though when it tries to build rabbit_common's jsx dependency.

We're trying to build from source since we want to roll our own custom authentication backend plugin

**Reply by lukebakken** [reply]
I'm not sure what's going on. I don't have an OS X machine to try and reproduce, either. There are RabbitMQ team members who use OS X without issue.

See this discussion, there's a chance you have a line-ending issue though I don't see evidence in the output you provided:

https://groups.google.com/g/rabbitmq-users/c/u9XT19TU3SU?pli=1

@michaelklishin @kjnilsson any suggestions for building on OS X? When I used to get the `undef` error for `rmemo:start` on OS X I would clean everything out to fix the issue. That doesn't appear to work here or I'm missing a step.

**Reply by ErikBZ** [reply]
I recloned the repository and now it seems to be building just fine. There may have been a file that was causing the issue that wasn't caught  by `git clean -fxd`. I had tried to build the project before I verified that I had all the dependencies installed. 

Another issue that may have contributed to this is that I also tried to build it via a VM through a shared folder. After I cloned it in on the VMs disk, it worked, which prompted me to try to reclone the project on my base OSX machine.

Thanks for your time!

**Reply by lukebakken** [reply]
Thanks for reporting back.

> There may have been a file that was causing the issue that wasn't caught by git clean -fxd

That's probably it. Still, ugh computers.

**Reply by lhoguin** [reply]
The `undef` error is related to a file in the `.erlang.mk` temporary directory. `make distclean` would normally remove it, but `git clean -xfffd` works just as well. I am not sure why you got an `undef` in the first place though.

**Comment by lukebakken** [comment]
FWIW, I also gave you bad instructions for cleaning your clone. These are the commands that should be run:

```
git clean -fxd
git pull origin master
```

Removing the `deps/` in entirety also removes sub directories that are part of the clone.
